src/Be.Vlaanderen.Basisregisters.AcmIdm/AuthorizationHandlers/AcmIdmAuthorizationHandler.cs

```diff
@@ -1,54 +1,61 @@
-﻿namespace Be.Vlaanderen.Basisregisters.AcmIdm.AuthorizationHandlers
+namespace Be.Vlaanderen.Basisregisters.AcmIdm.AuthorizationHandlers
 {
     using System;
-    using System.Collections.Generic;
-    using System.Collections.ObjectModel;
     using System.Linq;
     using System.Security.Claims;
     using System.Threading.Tasks;
     using Microsoft.AspNetCore.Authorization;
+    using NisCodeService.Abstractions;
 
     public class AcmIdmAuthorizationHandler : AuthorizationHandler<AcmIdmAuthorizationRequirement>
     {
-        private static readonly ReadOnlyDictionary<string, string> Cache = new ReadOnlyDictionary<string, string>(new Dictionary<string, string>(1000));
+        private readonly INisCodeService? _nisCodeService;
 
-        protected override Task HandleRequirementAsync(
+        public AcmIdmAuthorizationHandler(INisCodeService? nisCodeService = default)
+        {
+            _nisCodeService = nisCodeService;
+        }
+
+        protected override async Task HandleRequirementAsync(
             AuthorizationHandlerContext context,
             AcmIdmAuthorizationRequirement requirement)
         {
             if (requirement.AllowedValues.Any(scope => context.User.HasClaim(x => x.Type == AcmIdmClaimTypes.Scope && x.Value == scope)))
             {
-                AddNisCodeClaim(context);
+                await AddNisCodeClaim(context);
 
                 context.Succeed(requirement);
-                return Task.CompletedTask;
+                return;
             }
 
             context.Fail();
-            return Task.CompletedTask;
         }
 
-        private void AddNisCodeClaim(AuthorizationHandlerContext context)
+        private async Task AddNisCodeClaim(AuthorizationHandlerContext context)
         {
             var orgCodeClaim = context.User.Claims.SingleOrDefault(x => x.Type == AcmIdmClaimTypes.VoOrgCode);
             if (orgCodeClaim is null)
             {
                 return;
             }
 
-            var nisCodeClaim = Cache.ContainsKey(orgCodeClaim.Value)
-                ? Cache[orgCodeClaim.Value]
-                : FetchNisCode(orgCodeClaim.Value);
+            var nisCodeClaim = await FetchNisCode(orgCodeClaim.Value);
```

Comment:
Shouldn't we check if the NisCode claim is already present in the token? By adding multiple [Authorize] attributes on a controller action, the authorization handler is called multiple times (once per time the attribute is added).
![image](https://user-images.githubusercontent.com/18366653/211536840-1af378ef-1bfa-4562-9505-d8be25abdb0f.png)

Reply:
yes, will modify